applysolutionswriter.cpp

```diff
@@ -36,37 +38,47 @@ ApplySolutionsWriter::~ApplySolutionsWriter()
 
 void ApplySolutionsWriter::WriteBandInfo(const std::string &name, const std::vector<Writer::ChannelInfo> &channels, double refFreq, double totalBandwidth, bool flagRow)
 {
-	_nChannels = channels.size();
-	_correctedData.resize(_nChannels*4);
-	
+	_nBandFineChannels = channels.size();
+	_correctedData.resize(_nBandFineChannels*4);
+
 	ForwardingWriter::WriteBandInfo(name, channels, refFreq, totalBandwidth, flagRow);
 
-	if(_nChannels < _nSolutionChannels || (_nChannels % _nSolutionChannels) != 0) {
+	// We permit situations where the total number of fine channels is > solution channels but only when
+	// the solution channels can be divided into the total number of fine channels evenly.
+	if (_nSolutionChannels < _nTotalFineChannels || (_nSolutionChannels % _nTotalFineChannels) != 0) {
 		std::ostringstream s;
-		s << "The provided solution file has an incorrect number of channels. Oobservation has " << _nChannels << " channels, and solution file has " << _nSolutionChannels << " channels.";
+		s << "The provided solution file has an incorrect number of channels. Observation has " << _nTotalFineChannels << " channels, and solution file has " << _nSolutionChannels << " fine channels.";
 		throw std::runtime_error(s.str());
 	}
 }
 
-void ApplySolutionsWriter::WriteRow(double time, double timeCentroid, size_t antenna1, size_t antenna2, double u, double v, double w, double interval, const std::complex<float>* data, const bool* flags, const float* weights)
+void ApplySolutionsWriter::WriteRow(double time, double timeCentroid, size_t antenna1, size_t antenna2, double u, double v, double w, double interval, const std::complex<float> *data, const bool *flags, const float *weights)
 {
-	// Apply solution to averaged data
-	int channelRatio = _nChannels / _nSolutionChannels;
-	
+	// Apply solution to averaged data or where resolution of cal solution is higher than the obs you are applying it to
+	int channelRatio = _nSolutionChannels / _nTotalFineChannels;
```

Comment:
This looks incorrect to me, but I'm not sure... It used to be that there are at least as many channels in the data as that there are channels in the solutions, so _nTotalFineChannels >= _nSolutionChannels. In this case, channelRatio would always either round to 0 or to 1. This would then apply the wrong solution in the next lines of code...

Reply:
hmmm, I'm not sure now either! So, in my testing I was using examples where:

- Solution was produced from a 10 kHz, 20Khz, 40kHz calibrator obs, resulting in 3072,1536 or 768 channels in the solution file, but mostly the 10 kHz case as most of MWA calibrators since Phase 2 have 10 kHz fine channels.

- The observations had 3072,1536 or 768 channels if using -full-apply or 3072/N,1536/N or 768/N channels if using after averaging (-apply) where N is the averaging factor (2,4,...) e.g. averaging a 10 kHz fine channels to 80 kHz fine channels.

So this resulted in the vast majority of cases where I had more solution channels than observation/data channels and this required setting the ratio so we skip every N channels in the solution file (which I think is what you original code did too).

So this should result in channelRatio always being >=1. The case where the solution channels are less than the total fine channels is caught and prevented in the if statement check in `ApplySolutionsWriter::WriteBandInfo()`. 

Does this make sense? I could be wrong here- I'm still on a bit of a learning curve when it comes to calibration!

Reply:
Thanks Greg. I think the code never "skips" solution channels. For example, in `WriteBandInfo` it says:
```
	if(_nChannels < _nSolutionChannels || (_nChannels % _nSolutionChannels) != 0) {
		std::ostringstream s;
		s << "The provided solution file has an incorrect number of channels. Observation has " << _nChannels << " channels, and solution file has " << _nSolutionChannels << " channels.";
		throw std::runtime_error(s.str());
	}
```
So the number of channels in the solution file has to be at most the number of channels written, and therefore no solution channels are skipped. If you've tried to use a situation where nChan<nSolutions, you should have triggered that exception; didn't you get that exception? 

he code does allow the other way around: if there are fewer channels in the solution file than in the output, than a single solution channel is applied to multiple output channels. For that reason, the channelRatio was "output channels" / "solution channels", so always >= 1, and the solution channel was output channel / channelRatio, which would make sure e.g. with 100 output channels and 25 solution channels, that solution channel 0 is used for output channels 0-3, solution channel 1 for 4-7, etc. With your code, always channel 0 would be used, which I think is not correct.

Come to think of it, your code assumes the channels of all bands are in one solution file, but the original solution file that I defined was never used with an interrupted frequency axis. The calibrate tool will always calibrate a single band in picket fence observations, not the full band, so there should always be a solution file per band, and not one solution file with all bands as is assumed in your code. Of course it's possible to concatenate multiple bands into one solution file, but not sure that functionality exists already somewhere. Without something like that, I don't think it makes sense to apply solutions like this. I see two possible solutions you could pursue:
- Read a solution file per band in Cotter. It's a bit of a logistical hasle, but doable. 
- Make sure solution files can hold information for all bands (in picket fence observations). This is probably easier, but requires concatenating the results from multiple calibrate runs.

All in all, it's probably important that you try to test the following sequence: calibrate some picket fence obs -> apply those solutions to some picket fence obs with cotter -> image to see if things go correctly.

Reply:
Hi Andre, thanks for that.

@marcinsokolowski has produced a calibration database for pretty much all calibrator observations and @andreww5au built web services to allow anyone to download these (see [MWA Web Services on the wiki](https://wiki.mwatelescope.org/display/MP/Web+Services)). @marcinsokolowski  provides a single .bin file in the cases where the calibrator is picket fence.  The MWA ASVO uses this webservice, along with cotter to apply cal solutions (when requested) and produce a calibrated measurement set (or sets if picket fence), so the use case of all calibrator channels being in the one .bin file regardless is the main use case. I believe if people want to use cotter themselves with their own cal solutions, then there is trivial code which can combine many .bin with different channels into one .bin file, so I think this is ok. So, yeah we are already implmenting your second bullet point so it's good we are on the same page!

Just with: "I think the code never "skips" solution channels. For example, in WriteBandInfo it says:", I think the code you showed is your original code, but in my pull request I had changed it to:
```
if (_nSolutionChannels < _nTotalFineChannels || (_nSolutionChannels % _nTotalFineChannels) != 0) {
	std::ostringstream s;
	s << "The provided solution file has an incorrect number of channels. Observation has " << _nTotalFineChannels << " channels, and solution file has " << _nSolutionChannels << " fine channels.";
	throw std::runtime_error(s.str());
}
```
So if the number of channels in the solution file is less than the number of channels in the data, we will throw an exception. So for example if I try to apply a solution which has 768 channels to an observation of 1536 channels then this exception gets thrown.

hmm! So I thought at the time it was correct to disallow the condition of solution channels < obs channels, but it wouldn't be too hard to add to my change. What do you think?

I do believe that the case of solution chans >= observation chans works with the code I've supplied (and for picket fence cases)- I've done several cycles of cotter (-full-apply and -apply) then wsclean but I'm only a novice at radio imaging, but I know enough to tell the difference between an calibrated image and a calibrated one, especially for a point-like source. 

So, when comparing to an un-calibrated image the results are very encouraging- 4 example images attached non-calibrated and calibrated for obsid 1242552568 (ch057-068 and ch121-132 respectively) of HydA. I can provide the fits versions if you're keen to take a look, as the attached jpgs don't really work for this but you get the idea! In this case though, I got the cal solution file from John Morgan who concatenated the output of calibrate into a single .bin file for me as there have been issues getting the cal solution via the web services for this obs as the MRO has been offline for a few days with power / cooling issues.

I'll do some more testing over the next few days anyway to be more sure it works as expected.

1242552568_057-068_calibrated:
![1242552568_057-068_calibrated](https://user-images.githubusercontent.com/20233878/87646457-75d7b180-c781-11ea-9fa0-c87b49ae195a.jpg)

1242552568_057-068_not_calibrated:
![1242552568_057-068_not_calibrated](https://user-images.githubusercontent.com/20233878/87646463-7708de80-c781-11ea-9772-05c04f6edb96.jpg)

1242552568_121-132_calibrated:
![1242552568_121-132_calibrated](https://user-images.githubusercontent.com/20233878/87646467-77a17500-c781-11ea-9e29-7d2b30263d06.jpg)

1242552568_121-132_not_calibrated:
![1242552568_121-132_not_calibrated](https://user-images.githubusercontent.com/20233878/87646471-78d2a200-c781-11ea-85a1-046e8eb52334.jpg)

Reply:
Hi Greg,

You'll probably have to do some deeper imaging and differential comparison to notice the difference of applying the wrong channel, but I think things will only go wrong when the solution file has fewer solution than the data, and in that case you say you throw an exception, so then the solutions will never be incorrectly applied.

However, having fewer solution channels than in the data is a reasonably common use-case, as one can do this sequence:

- Run Cotter on a calibrator data and average in frequency (e.g. 4 times)
- Calibrate the averaged data, produce a solution file with Nchan/4 channels.
- Run Cotter on target data and apply the solutions

In this case, the solutions need to be "extrapolated", i.e., a single solution channel needs to be applied to multiple data channels. I don't know how common this is but I think it should be fairly simple to keep supporting this. The case you're now supporting is "run Cotter without averaging, calibrate, run Cotter with averaging and apply solutions afterwards". This seems to be a strange use-case to me, and would already be supported by Cotter by applying the full-res solutions before averaging.

Reply:
Hi @aroffringa, I've updated my local copy of the code to support both scenarios (well 3 scenarios)-
solution channels > data channels (before or after averaging)
solution channels < data channels (before or after averaging)
solution channels == data channels (before or after averaging)

My use case was that Marcin's web service produces me a calibration solution, and then I run cotter and apply it before averaging (this is the MWA ASVO use case), so hopefully not so strange when explained this way?  :-)

I'm currently doing some tests by imaging these various scenarios and ensuring the solutions are applied correctly in all cases. Once I'm pretty sure of that I'll push my commit and you can take a look!